Fix reindex duration bug when bag file durations overlap

ChuiVanfleet opened the pull request:
## Description

If bag file start and end times overlap with each other, then the duration of the re-indexed bag is wrong.
For example if I have two bag files that start at the same time, and each is 20seconds long, then the reindexer will say that the duration is 40 seconds instead of 20 seconds.

### Is this user-facing behavior change?

Not sure if this counts, but the duration field in the reindexed metadat.yaml file will be correct.

### Did you use Generative AI?

No.

### Additional Information

This is my first PR back to ros2-related repositories, so I'm not sure if I'm doing this right!

Commit:
Fix reindex duration bug when bag files overlap.

Signed-off-by: Chui Vanfleet <chui.vanfleet@epiroc.com>

Commit:
Set the duration instead of adding it.

Signed-off-by: Chui Vanfleet <chui.vanfleet@epiroc.com>

Commit:
Fixing some linting.

Signed-off-by: Chui Vanfleet <chui.vanfleet@epiroc.com>

MichaelOrlov committed:
Cleanups in test_reindexer.cpp

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov reviewed (approved):
@ChuiVanfleet Thank you for your contribution.

The failed scenario described in the description to this PR shouldn't have happened in the standard workflow, because bags in the recorded folder shouldn't overlap by time **by design**.  
The described failure could happen only if one manually mixed and matched recorded files from different recorder instances or something similar.

In general, I don't mind the proposed fix. It does make sense to me.  
However, I had a few concerns about how the unit test was written. I've made a bit of  clean-ups there. Please refer to my commit https://github.com/ros2/rosbag2/pull/2036/commits/e43a22d1afc511cf8b8dbfde55b0a9df7ea75142. 

Now looks good to me with green CI.

MichaelOrlov commented:
Pulls: ros2/rosbag2#2036
Gist: https://gist.githubusercontent.com/MichaelOrlov/146a3767b90d1a78249df1fa78e908e9/raw/842275466d1301e9218c2c6ffd0327822ad293e1/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_cpp ros2bag rosbag2_tests
TEST args:  --packages-above rosbag2_cpp ros2bag rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16642
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24502)](http://ci.ros2.org/job/ci_linux/24502/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18558)](http://ci.ros2.org/job/ci_linux-aarch64/18558/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=3963)](http://ci.ros2.org/job/ci_linux-rhel/3963/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25055)](http://ci.ros2.org/job/ci_windows/25055/)

MichaelOrlov commented:
@ros-pull-request-builder retest this please

MichaelOrlov committed:
Attempt to fix Windows build error

Fix for error:
 "test_reindexer.cpp(50,1): error C2679: binary '=': no operator found
 which takes a right-hand operand of type
 'std::chrono::system_clock::time_point' (or there is no acceptable
 conversion)

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov commented:
Re-run CI after an attempt to fix the Windows build error
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16648
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24511)](http://ci.ros2.org/job/ci_linux/24511/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18564)](http://ci.ros2.org/job/ci_linux-aarch64/18564/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=3972)](http://ci.ros2.org/job/ci_linux-rhel/3972/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25063)](http://ci.ros2.org/job/ci_windows/25063/)

fujitatomoya reviewed (approved):
lgtm!

MichaelOrlov commented:
https://github.com/Mergifyio backport kilted jazzy
